src/compiler/transformers/module/module.ts

```diff
@@ -88,11 +88,14 @@ namespace ts {
             addRange(statements, endLexicalEnvironment());
             addExportEqualsIfNeeded(statements, /*emitAsReturn*/ false);
 
+            if (!currentModuleInfo.exportEquals) {
+                append(statements, createUnderscoreUnderscoreESModule());

```

Comment:
@yuit not to *prevent* circular imports, but rather to *enable* them. Consider:

```ts
// a.ts
import b from './b';
export default function(x) {
    return x ? b(false) : 1;
}

// b.ts
import a from './a';
export default function(x) {
    return x ? a(false) : 2;
}
```

Both `a.ts` and `b.ts` **should** be recognized as an ES module as early as possible, before the first `require()` is called in either module.

Consider the following scenario, where we emit `__esModule` at the end:

```js
// a.js
var b_1 = require("./b");
exports.default = function (x) {
    return x ? b_1(false) : 1;
};
Object.defineProperty(exports, "__esModule", { value: true });

// b.ts
var a_1 = require("./a");
exports.default = function (x) {
    return x ? a_1(false) : 2;
}
Object.defineProperty(exports, "__esModule", { value: true });
```

Assuming `a.js` is loaded first:

1. A module object will be created for `a.js` and stored in Node's module cache. At this point, Node (or rather, some future version) does not know whether it is an ES module or a CJS module.  
2. The `a.js` module body begins to execute and `a.js` invokes a `require()` for `b.js`.  
3. A module object will be created for `b.js` and stored in Node's module cache.  
4. The `b.js` module body begins to execute the module body and invokes a `require()` for `a.js`.   
5. Node cannot see the `__esModule` marker on the exports object for `a.js` and doesn't know is an ES module, and therefore cannot provide the right answer to `b.js`.  

Now consider the same scenario, with the `__esModule` marker at the beginning:

```js
// a.js
Object.defineProperty(exports, "__esModule", { value: true });
var b_1 = require("./b");
exports.default = function (x) {
    return x ? b_1(false) : 1;
};

// b.ts
Object.defineProperty(exports, "__esModule", { value: true });
var a_1 = require("./a");
exports.default = function (x) {
    return x ? a_1(false) : 2;
}
```

Assuming `a.js` is loaded first:

1. A module object will be created for `a.js` and stored in Node's module cache. At this point, Node (or rather, some future version) does not know whether it is an ES module or a CJS module.  
2. The `a.js` module body begins to execute and `a.js` invokes a `require()` for `b.js`.  
3. A module object will be created for `b.js` and stored in Node's module cache.  
4. The `b.js` module body begins to execute the module body and invokes a `require()` for `a.js`.  
5. Node can see the `__esModule` marker on the exports object for `a.js` and knows it can treat it as an ES module.  

However, `export =` cannot be used circularly to begin with so this whole topic is possibly moot.